File: python/src/fdkey/jwt_verify.py
"""Ed25519 JWT verification using PyJWT[crypto]. Same wire shape as TS SDK."""

from __future__ import annotations

from typing import Any, Optional

import jwt as pyjwt

from .well_known import WellKnownClient

CLOCK_TOLERANCE_SECONDS = 30


class JwtVerifier:
    """Wraps WellKnownClient and pyjwt to verify FDKEY-issued tokens.

    Returns the decoded claims dict on success, None on any failure
    (missing kid, unknown kid, bad signature, expired). Caller treats
    None as 'verification failed' and surfaces accordingly."""

    def __init__(self, well_known: WellKnownClient) -> None:
        self._well_known = well_known

    async def verify(self, token: str) -> Optional[dict[str, Any]]:
        try:
            header = pyjwt.get_unverified_header(token)
        except Exception:
            return None
        kid = header.get("kid")
        if not kid:
            return None
        key = await self._well_known.get_key(kid)
        if key is None:
            return None
        try:
            claims = pyjwt.decode(
                token,
                key=key,
                algorithms=["EdDSA"],
                leeway=CLOCK_TOLERANCE_SECONDS,
                # FDKEY's audience is the integrator's vps_users.id, which
                # the SDK doesn't know at verify time. We accept any aud
                # for now (the VPS already binds aud to the api_key that
                # requested the challenge — defense in depth).
                options={"verify_aud": False},
            )
        except Exception:
            return None
        if not isinstance(claims, dict):
            return None
        return claims


def extract_score(claims: Optional[dict[str, Any]]) -> Optional[float]:
    if not claims:
        return None
    v = claims.get("score")
    if isinstance(v, (int, float)):
        return float(v)
    return None


def extract_tier(claims: Optional[dict[str, Any]]) -> Optional[str]:
    if not claims:
        return None
    v = claims.get("tier")
    if isinstance(v, str):
        return v
    return None

File: rust/src/jwt_verify.rs
//! Ed25519 JWT verification. Same wire shape as the TypeScript SDK.

use std::collections::HashMap;

use jsonwebtoken::{decode, decode_header, Algorithm, Validation};
use serde::Deserialize;

use crate::types::FdkeyError;
use crate::well_known::WellKnownClient;

const CLOCK_TOLERANCE_SECONDS: u64 = 30;

#[derive(Debug, Deserialize)]
pub struct VerifiedJwt {
    pub score: f64,
    pub tier: String,
    #[serde(default)]
    pub puzzle_summary: serde_json::Value,
    #[serde(flatten)]
    pub extra: HashMap<String, serde_json::Value>,
}

pub struct JwtVerifier {
    well_known: WellKnownClient,
}

impl JwtVerifier {
    pub fn new(well_known: WellKnownClient) -> Self {
        Self { well_known }
    }

    /// Verify a Bearer JWT and return the decoded claims. Errors on any
    /// failure — bad signature, expired (modulo 30s tolerance), unknown kid.
    pub async fn verify(&self, token: &str) -> Result<VerifiedJwt, FdkeyError> {
        let header = decode_header(token)?;
        let kid = header.kid.ok_or(FdkeyError::MissingKid)?;
        let key = self
            .well_known
            .get_key(&kid)
            .await?
            .ok_or_else(|| FdkeyError::UnknownKid(kid.clone()))?;

        let mut validation = Validation::new(Algorithm::EdDSA);
        validation.leeway = CLOCK_TOLERANCE_SECONDS;
        // FDKEY's audience is the integrator's vps_users.id, which the
        // SDK doesn't know at verify time. Defense-in-depth: the VPS
        // already binds aud to the api_key that requested the challenge.
        validation.validate_aud = false;

        let data = decode::<serde_json::Value>(token, &key, &validation)?;
        let claims = serde_json::from_value::<VerifiedJwt>(data.claims)?;
        Ok(claims)
    }
}

/// Pull a Bearer token out of an `Authorization` header value. Returns None
/// for missing, empty, malformed, or non-`Bearer` schemes. Case-insensitive.
pub fn extract_bearer(header_value: Option<&str>) -> Option<&str> {
    let value = header_value?.trim();
    let (scheme, token) = value.split_once(char::is_whitespace)?;
    if !scheme.eq_ignore_ascii_case("bearer") {
        return None;
    }
    let token = token.trim();
    if token.is_empty() {
        None
    } else {
        Some(token)
    }
}
